syntax = "proto2";
package tzrec.protos;

message AUC {
    optional uint32 thresholds = 1 [default = 200];
}

message MulticlassAUC {
    optional uint32 thresholds = 1 [default = 200];
    // macro: calculate score for each class and average them
    // weighted: calculates score for each class and computes weighted average using their support
    optional string average = 2 [default = 'macro'];
}

message RecallAtK {
    optional uint32 top_k = 1 [default = 5];
}

message MeanAbsoluteError {
}

message MeanSquaredError {
}

message Accuracy {
    // probs threshold
    optional float threshold = 1 [default = 0.5];
    // top k accuracy when num_class > 1
    optional uint32 top_k = 2 [default = 1];
}

message GroupedAUC {
    required string grouping_key = 1;
}

message XAUC {
    optional float sample_ratio = 1 [default = 1e-3];
    optional uint64 max_pairs = 2;
    optional bool in_batch = 3 [default = false];
}

message GroupedXAUC {
    required string grouping_key = 1;
    optional uint64 max_pairs_per_group = 2 [default = 100];

}

message MetricConfig {
    oneof metric {
        AUC auc = 1;
        MulticlassAUC multiclass_auc = 2;
        RecallAtK recall_at_k = 3;
        MeanAbsoluteError mean_absolute_error = 4;
        MeanSquaredError mean_squared_error = 5;
        Accuracy accuracy = 6;
        GroupedAUC grouped_auc = 7;
        XAUC xauc = 8;
        GroupedXAUC grouped_xauc = 9;
    }
}
